# Data Science Aptitude Test - Daimler Group

- date: yyyy.MM.dd
- author: [name_user](https://www.linkedin.com/in/)
***

## 1. Background 

The following test consists in one scenario in which you will have to analyze and train a model for one data set. The data set contains information about year, price, transmission, mileage, fuel type and engine size of used C-class cars. The idea is training a model to **predict** which will be the price of a used *C-class car* in the market.

### Data Science Workflow stages

The competition solution workflow goes through ten stages described in the Data Science Solutions book.

1. Check Environment & libraries version.
2. Import Libraries.
3. Question or problem definition.
4. Acquire dataset or training and testing data.
5. Analyze, identify patterns, and explore the data.
6. Wrangle, prepare, cleanse the data.
7. Feature Engineering.
8. Model, predict and MLOps.
9. Visualize, report, and present the problem solving steps and final solution.
10. Create or solve AI pipeline on Cloud, architecture solution and shows results.

Could be to exist another substeps during Data Processing and Data Modeling, for to train and deploy many machine learning and/or deep learning combination pipelines.

## 2. This notebook covers only:

1. [X] Upload the csv file to your workspace and load it into a data frame.
2. [X] Look for null values and outliers. Remove, keep or impute them and explain why you did so.
3. [X] Show the main statistics (mean, standard deviation…) of the numerical columns of the data set. Are any of the variables skewed? (You can use any visualization you need to answer the last question).
4.	Train a model for the prediction of the price. Explain why you chose the model that you have trained.
5.	Test the model and obtain some performance metrics from it. Would you say that the model has a good performance? Why?
6.	Would you say that you have enough information to predict the price of an EQC (electric C-class)? Why?

## 3. The business problem

To be able to predict used cars market value can help both buyers and sellers.

Used car sellers (dealers): They are one of the biggest target group that can be interested in results of this study. If used car sellers better understand what makes a car desirable, what the important features are for a used car, then they may consider this knowledge and offer a better service.

Online pricing services: There are websites that offers an estimate value of a car. They may have a good prediction model. However, having a second model may help them to give a better prediction to their users. Therefore, the model developed in this study may help online web services that tells a used car’s market value.

Individuals: There are lots of individuals who are interested in the used car market at some points in their life because they wanted to sell their car or buy a used car. In this process, it’s a big corner to pay too much or sell less then it’s market value.

What determines the price of used cars? 

<img src="https://s1.dmcdn.net/v/L5C9Y1QCcGAoct-M5/x1080" width="200">


The value of a car drops right from the moment it is bought and the depreciation continues with each passing year. In fact, in the first year itself, the value of a car decreases by 20 percent of its initial value. The make and model of a car, total kilometers driven, overall condition of the vehicle and various other factors further affect the car’s resale value .

Our results show that Random Forest model and K-Means clustering with linear regression yield the best results, but are compute heavy.
Conventional linear regression also yielded satisfactory results, with the advantage of a significantly lower training time in comparison to the aforementioned methods. We check XGBoost, Linear Regression and Decision Tree estimators.

## 4. ETL (Extract Transform Load)

Loading dataset from csv file. Anyway in other notebook in attachment, we will be able to scrape specialized website and get other important features.

In [0]:
# File location and type cclass dataset stored on Databricks


# CSV options


# The applied options are for CSV files. For other file types, these will be ignored.


model year price transmission mileage fuelType engineSize C Class 2020 30495 Automatic 1200 Diesel 2.0 C Class 2020 29989 Automatic 1000 Petrol 1.5 C Class 2020 37899 Automatic 500 Diesel 2.0 C Class 2019 30399 Automatic 5000 Diesel 2.0 C Class 2019 29899 Automatic 4500 Diesel 2.0 C Class 2020 30999 Automatic null Diesel 2.0 C Class 2020 35999 Automatic 500 Diesel 2.0 C Class 2019 37990 Automatic 1412 Petrol 3.0 C Class 2019 28990 Automatic 3569 Diesel 2.0 C Class 2019 28990 Automatic 3635 Diesel 2.0 C Class 2013 9995 Automatic 44900 Petrol 1.6 C Class 2012 6995 Automatic 88200 Diesel 2.1 C Class 2012 7495 Automatic 115000 Diesel 2.1 C Class 2011 8995 Automatic 69250 Diesel 2.1 C Class 2015 14995 Automatic 49850 Diesel 2.1 C Class 2013 8595 Automatic 82685 Diesel 2.1 C Class 2016 19250 Semi-Auto 32506 Diesel 2.1 C Class 2017 23500 Semi-Auto 17000 Diesel 2.1 C Class 2018 24995 Semi-Auto 17744 Diesel 2.0 C Class 2016 29995 Semi-Auto 64291 Petrol 4.0 C Class 2014 17990 Automatic 41011 Petrol 2.0 C Class null 14477 Automatic 72000 Diesel 1.6 C Class 2013 10977 Automatic 72000 Diesel 2.1 C Class 2016 17977 Automatic 40000 Diesel 1.6 C Class 2015 12790 Automatic 52000 Diesel 2.1 C Class 2017 14488 Automatic 66472 Diesel 2.1 C Class 2011 7500 Manual 91000 Diesel 2.1 C Class 2013 9750 Automatic 69700 Diesel 2.1 C Class 2015 15995 Automatic 39000 Diesel 1.6 C Class 2016 15675 Manual 39287 Diesel 2.1 C Class 2016 16475 Manual 40569 Diesel 2.1 C Class 2016 15975 Manual 37419 Petrol 2.0 C Class 2016 16475 Automatic 34452 Diesel 2.1 C Class 2015 15975 Automatic 41567 Hybrid 2.1 C Class 2015 16495 Automatic null Hybrid 2.0 C Class 2017 17795 Automatic 37778 Diesel 2.1 C Class 2017 17875 Automatic 44325 Diesel 2.1 C Class 2016 14000 Automatic 45000 Diesel 2.1 C Class 2017 16500 Automatic 75000 Diesel 1.6 C Class 2016 21000 Automatic 36000 Diesel 2.1 C Class 2015 15500 Automatic 64000 Diesel 2.1 C Class 2013 7250 Manual 106000 Diesel 2.1 C Class 2015 13000 Automatic 67150 Diesel 2.1 C Class 2012 11000 Automatic 43426 Petrol 1.6 C Class 2016 13300 Manual 42622 Diesel 2.1 C Class 2016 16000 Manual 34903 Diesel 2.1 C Class 2014 12500 Automatic 35345 Diesel 2.1 C Class 2015 14495 Manual 23146 Petrol 2.0 C Class 2017 24500 Automatic 16000 Diesel 2.1 C Class 2013 23990 Automatic 44000 Petrol 6.2 C Class 2015 17250 Automatic 30000 Diesel 2.1 C Class 2015 14200 Automatic 22000 Petrol 1.6 C Class 2017 22950 Automatic 19000 Petrol 2.0 C Class 2013 10500 Automatic 90000 Diesel 2.1 C Class 2014 15650 Automatic 32000 Diesel 2.1 C Class 2015 14850 Automatic 90000 Diesel 2.1 C Class 2017 22400 Automatic 37000 Diesel 2.1 C Class 2017 19950 Automatic 23000 Diesel 2.1 C Class 2014 9990 Automatic 78000 Diesel 2.1 C Class 2015 11990 Automatic 47000 Diesel 2.1 C Class 2017 20990 Automatic 18000 Diesel 2.1 C Class 2015 14490 Automatic 64000 Diesel 2.1 C Class 2013 10490 Automatic 50000 Diesel 2.1 C Class 2016 15990 Automatic 65000 Diesel 2.1 C Class 2016 17850 Automatic 38000 Diesel 2.1 C Class 2016 16450 Automatic 70000 Diesel 2.1 C Class 2013 11350 Automatic 69000 Diesel 2.1 C Class 2014 12900 Automatic 41000 Petrol 1.6 C Class 2018 26000 Automatic 16250 Petrol 1.5 C Class 2017 22767 Automatic 30676 Diesel 2.1 C Class 2015 12000 Automatic 40005 Diesel 2.1 C Class 2017 18999 Automatic 38000 Hybrid 2.0 C Class 2016 14000 Automatic 45000 Diesel 2.1 C Class 2015 12000 Automatic 40005 Diesel 2.1 C Class 2009 4343 Automatic 120000 Diesel 2.1 C Class 2016 16500 Automatic 51936 Diesel 2.1 C Class 2019 27750 Automatic 4000 Diesel 2.0 C Class 2019 28950 Automatic 8500 Diesel 2.0 C Class 2007 3495 Automatic 119168 Petrol 1.8 C Class 2017 59740 Automatic 21800 Petrol 4.0 C Class 2019 44990 Automatic 19999 Petrol 4.0 C Class 2019 54990 Automatic 1600 Petrol 4.0 C Class 2015 15995 Automatic 46221 Diesel 2.1 C Class 2016 14000 Automatic 45000 Diesel 2.1 C Class 2015 12000 Automatic 40005 Diesel 2.1 C Class 2017 22767 Automatic 30676 Diesel 2.1 C Class 2014 12998 Semi-Auto 5271

In [0]:
# Create a custom schema to pass data types correctly.
from pyspark.sql.types import *


In [0]:
# Load dataset with custom schema


# The applied options are for CSV files. For other file types, these will be ignored.


In [0]:
# Check basic statistics like mean, standard deviation, min and max
display(df.describe(["price","mileage"]))

summary price mileage count 3899 3892 mean 23674.28699666581 22396.172661870503 stddev 8960.21821842348 22641.50657329362 min 1290 1 max 88995 173000

> Check distribution later.

In [0]:
# Create a view or temporary table to manipulate in SparkSQL



In [0]:
%sql

/* Query the created temp table in a SQL cell, show all features limited 25 rows only */



model year price transmission mileage fuelType engineSize C Class 2020-01-01 30495 Automatic 1200 Diesel 2.0 C Class 2020-01-01 29989 Automatic 1000 Petrol 1.5 C Class 2020-01-01 37899 Automatic 500 Diesel 2.0 C Class 2019-01-01 30399 Automatic 5000 Diesel 2.0 C Class 2019-01-01 29899 Automatic 4500 Diesel 2.0 C Class 2020-01-01 30999 Automatic null Diesel 2.0 C Class 2020-01-01 35999 Automatic 500 Diesel 2.0 C Class 2019-01-01 37990 Automatic 1412 Petrol 3.0 C Class 2019-01-01 28990 Automatic 3569 Diesel 2.0 C Class 2019-01-01 28990 Automatic 3635 Diesel 2.0 C Class 2013-01-01 9995 Automatic 44900 Petrol 1.6 C Class 2012-01-01 6995 Automatic 88200 Diesel 2.1 C Class 2012-01-01 7495 Automatic 115000 Diesel 2.1 C Class 2011-01-01 8995 Automatic 69250 Diesel 2.1 C Class 2015-01-01 14995 Automatic 49850 Diesel 2.1 C Class 2013-01-01 8595 Automatic 82685 Diesel 2.1 C Class 2016-01-01 19250 Semi-Auto 32506 Diesel 2.1 C Class 2017-01-01 23500 Semi-Auto 17000 Diesel 2.1 C Class 2018-01-01 24995 Semi-Auto 17744 Diesel 2.0 C Class 2016-01-01 29995 Semi-Auto 64291 Petrol 4.0 C Class 2014-01-01 17990 Automatic 41011 Petrol 2.0 C Class null 14477 Automatic 72000 Diesel 1.6 C Class 2013-01-01 10977 Automatic 72000 Diesel 2.1 C Class 2016-01-01 17977 Automatic 40000 Diesel 1.6 C Class 2015-01-01 12790 Automatic 52000 Diesel 2.1

## 5. Data Visualization

The Exploratory Data Analysis (EDA)

While exploring the data, we will look at the different combinations of features with the help of visuals. This will help us to understand our data better and give us some clue about pattern in data.

In [0]:
# Validating data


root
-- model: string (nullable = true)
-- year: date (nullable = true)
-- price: integer (nullable = true)
-- transmission: string (nullable = true)
-- mileage: integer (nullable = true)
-- fuelType: string (nullable = true)
-- engineSize: string (nullable = true)

In [0]:
# Dimension of Dataset


Dataset has 3899 observations and has 7 features.

In [0]:
# Show 25 sample values


Out[8]:

model 
 year 
 price 
 transmission 
 mileage 
 fuelType 
 engineSize 
 
 
 
 
 403 
 C Class 
 2019-01-01 
 31980 
 Semi-Auto 
 9.0 
 Petrol 
 1.5 
 
 
 990 
 C Class 
 2019-01-01 
 29299 
 Semi-Auto 
 2758.0 
 Petrol 
 2.0 
 
 
 1100 
 C Class 
 2019-01-01 
 20349 
 Semi-Auto 
 13910.0 
 Petrol 
 1.5 
 
 
 935 
 C Class 
 2019-01-01 
 20599 
 Semi-Auto 
 12091.0 
 Petrol 
 1.5 
 
 
 883 
 C Class 
 2018-01-01 
 43750 
 Semi-Auto 
 11938.0 
 Petrol 
 4.0 
 
 
 598 
 C Class 
 2017-01-01 
 22162 
 Semi-Auto 
 48142.0 
 Diesel 
 2.1 
 
 
 573 
 C Class 
 2018-01-01 
 24546 
 Automatic 
 12397.0 
 Diesel 
 2.1 
 
 
 1813 
 C Class 
 2015-01-01 
 15300 
 Automatic 
 33303.0 
 Diesel 
 2.1 
 
 
 659 
 C Class 
 2019-01-01 
 19980 
 Semi-Auto 
 13936.0 
 Petrol 
 1.5 
 
 
 761 
 C Class 
 2019-01-01 
 31495 
 Automatic 
 3500.0 
 Diesel 
 2.0 
 
 
 1917 
 C Class 
 2012-01-01 
 11995 
 Other 
 44644.0 
 Diesel 
 3.0 
 
 
 1249 
 C Class 
 2018-01-01 
 29499 
 Automatic 
 1988.0 
 Petrol 
 2.0 
 
 
 435 
 C Class 
 2017-01-01 
 20995 
 Semi-Auto 
 19750.0 
 Diesel 
 2.1 
 
 
 315 
 C Class 
 2014-01-01 
 13251 
 Semi-Auto 
 56219.0 
 Diesel 
 2.1 
 
 
 889 
 C Class 
 2016-01-01 
 20895 
 Semi-Auto 
 35999.0 
 Diesel 
 2.1 
 
 
 323 
 C Class 
 2016-01-01 
 14700 
 Automatic 
 59023.0 
 Diesel 
 1.6 
 
 
 1805 
 C Class 
 2017-01-01 
 16600 
 Automatic 
 45277.0 
 Diesel 
 2.1 
 
 
 872 
 C Class 
 2019-01-01 
 25995 
 Automatic 
 4235.0 
 Diesel 
 1.6 
 
 
 1732 
 C Class 
 2016-01-01 
 17799 
 Automatic 
 53613.0 
 Diesel 
 2.1 
 
 
 236 
 C Class 
 2017-01-01 
 16500 
 Semi-Auto 
 14077.0 
 Diesel 
 2.1 
 
 
 405 
 C Class 
 2015-01-01 
 15495 
 Automatic 
 37200.0 
 Diesel 
 2.1 
 
 
 492 
 C Class 
 2018-01-01 
 23500 
 Automatic 
 4368.0 
 Petrol 
 1.5 
 
 
 1106 
 C Class 
 2019-01-01 
 25899 
 Semi-Auto 
 6765.0 
 Petrol 
 1.5 
 
 
 1863 
 C Class 
 2016-01-01 
 14000 
 Automatic 
 45000.0 
 Diesel 
 2.1 
 
 
 42 
 C Class 
 2016-01-01 
 16998 
 Semi-Auto 
 26973.0 
 Diesel 
 1.6

In [0]:
%sql


price fuelType 30495 Diesel 29989 Petrol 37899 Diesel 30399 Diesel 29899 Diesel 30999 Diesel 35999 Diesel 37990 Petrol 28990 Diesel 28990 Diesel 9995 Petrol 6995 Diesel 7495 Diesel 8995 Diesel 14995 Diesel 8595 Diesel 19250 Diesel 23500 Diesel 24995 Diesel 29995 Petrol 17990 Petrol 14477 Diesel 10977 Diesel 17977 Diesel 12790 Diesel 14488 Diesel 7500 Diesel 9750 Diesel 15995 Diesel 15675 Diesel 16475 Diesel 15975 Petrol 16475 Diesel 15975 Hybrid 16495 Hybrid 17795 Diesel 17875 Diesel 14000 Diesel 16500 Diesel 21000 Diesel 15500 Diesel 7250 Diesel 13000 Diesel 11000 Petrol 13300 Diesel 16000 Diesel 12500 Diesel 14495 Petrol 24500 Diesel 23990 Petrol 17250 Diesel 14200 Petrol 22950 Petrol 10500 Diesel 15650 Diesel 14850 Diesel 22400 Diesel 19950 Diesel 9990 Diesel 11990 Diesel 20990 Diesel 14490 Diesel 10490 Diesel 15990 Diesel 17850 Diesel 16450 Diesel 11350 Diesel 12900 Petrol 26000 Petrol 22767 Diesel 12000 Diesel 18999 Hybrid 14000 Diesel 12000 Diesel 4343 Diesel 16500 Diesel 27750 Diesel 28950 Diesel 3495 Petrol 59740 Petrol 44990 Petrol 54990 Petrol 15995 Diesel 14000 Diesel 12000 Diesel 22767 Diesel 12998 Diesel 17998 Diesel 28998 Petrol 24998 Petrol 19498 Diesel 23498 Diesel 25498 Diesel 24998 Diesel 18498 Hybrid 15498 Diesel 20995 Diesel 31995 Petrol 23995 Diesel 16995 Diesel 24995 Petrol 26998 Petrol 20498 Diesel 27498 Diesel 16998 Diesel 27498 Diesel 27698 Diesel 27498 Petrol 32498 Diesel 24498 Diesel 21498 Diesel 18998 Diesel 15998 Diesel 9298 Diesel 22998 Diesel 12998 Diesel 22998 Diesel 17998 Diesel 28498 Diesel 18998 Diesel 20998 Diesel 21998 Diesel 13998 Diesel 24498 Diesel 20998 Hybrid 30498 Diesel 19498 Diesel 15298 Diesel 33998 Petrol 32998 Petrol 16498 Diesel 15498 Diesel 15498 Diesel 18998 Diesel 24495 Diesel 18990 Diesel 10995 Diesel 15698 Diesel 22498 Petrol 19298 Diesel 16498 Hybrid 19998 Petrol 20685 Diesel 20489 Diesel 13800 Diesel 31998 Petrol 29998 Petrol 24998 Petrol 32998 Petrol 32998 Petrol 29998 Petrol 23998 Diesel 20998 Hybrid 16498 Diesel 11998 Diesel 12495 Diesel 19298 Petrol 23998 Diesel 12495 Diesel 17498 Hybrid 16198 Diesel 18498 Diesel 24998 Petrol 24998 Petrol 21998 Petrol 29498 Diesel 29498 Diesel 16998 Diesel 33998 Petrol 32998 Petrol 18998 Diesel 32498 Petrol 25798 Diesel 20498 Diesel 18498 Diesel 14498 Petrol 29498 Diesel 17998 Petrol 13498 Diesel 19698 Hybrid 22498 Hybrid 11998 Diesel 11998 Petrol 39998 Hybrid 23498 Diesel 57998 Petrol 16315 Diesel 13991 Diesel 16998 Diesel 16498 Diesel 16998 Diesel 13690 Diesel 14490 Diesel 17998 Petrol 29998 Diesel 18998 Hybrid 18998 Petrol 28998 Diesel 24998 Petrol 27998 Diesel 16798 Diesel 19998 Diesel 29950 Diesel 20790 Hybrid 9490 Petrol 15690 Diesel 14990 Diesel 17290 Diesel 19490 Diesel 28990 Diesel 24990 Diesel 28990 Petrol 16298 Diesel 14698 Diesel 13998 Petrol 14698 Diesel 10498 Diesel 11998 Diesel 15698 Petrol 16998 Diesel 17998 Hybrid 20998 Diesel 13691 Diesel 13491 Diesel 21998 Petrol 14500 Diesel 15990 Diesel 20490 Petrol 20890 Petrol 14490 Diesel 21990 Petrol 14750 Diesel 44990 Petrol 24690 Petrol 18270 Diesel 26625 Diesel 13361 Petrol 17490 Diesel 26427 Diesel 29290 Diesel 23047 Diesel 13621 Diesel 18200 Diesel 16991 Diesel 23498 Petrol 14698 Diesel 9798 Diesel 17998 Diesel 18998 Diesel 20700 Diesel 16990 Diesel 11791 Diesel 18191 Diesel 20991 Diesel 13321 Diesel 21998 Diesel 15998 Hybrid 13498 Diesel 20871 Diesel 25891 Petrol 20591 Diesel 19991 Diesel 15990 Diesel 30055 Diesel 21650 Petrol 25990 Petrol 17690 Diesel 30950 Diesel 24950 Petrol 14998 Diesel 19798 Hybrid 24498 Petrol 19991 Diesel 21991 Diesel 22591 Diesel 18790 Diesel 25950 Petrol 29950 Diesel 14990 Diesel 31950 Diesel 20990 Diesel 28990 Petrol 27000 Diesel 12590 Diesel 20885 Diesel 25885 Petrol 14965 Diesel 16595 Diesel 17940 Diesel 19940 Diesel 16498 Diesel 14500 Diesel 27791 Diesel 26499 Diesel 31990 Diesel 20999 Petrol 31999 Diesel 17495 Petrol 25495 Petrol 36995 Diesel 30999 Diesel 29999 Diesel 17880 Diesel 13660 Diesel 23920 Diesel 22960 Diesel 16590 Diesel 17990 Di

> First part of multivariate analysis, price is **right skewed** for most of fuelType feature, most frequent for Diesel and Petrol. Right-skewed distributions are also called positive-skew distributions. That’s because there is a long tail in the positive direction on the number line. The mean is also to the right of the peak.

In [0]:
%sql


mileage fuelType 1200 Diesel 1000 Petrol 500 Diesel 5000 Diesel 4500 Diesel null Diesel 500 Diesel 1412 Petrol 3569 Diesel 3635 Diesel 44900 Petrol 88200 Diesel 115000 Diesel 69250 Diesel 49850 Diesel 82685 Diesel 32506 Diesel 17000 Diesel 17744 Diesel 64291 Petrol 41011 Petrol 72000 Diesel 72000 Diesel 40000 Diesel 52000 Diesel 66472 Diesel 91000 Diesel 69700 Diesel 39000 Diesel 39287 Diesel 40569 Diesel 37419 Petrol 34452 Diesel 41567 Hybrid null Hybrid 37778 Diesel 44325 Diesel 45000 Diesel 75000 Diesel 36000 Diesel 64000 Diesel 106000 Diesel 67150 Diesel 43426 Petrol 42622 Diesel 34903 Diesel 35345 Diesel 23146 Petrol 16000 Diesel 44000 Petrol 30000 Diesel 22000 Petrol 19000 Petrol 90000 Diesel 32000 Diesel 90000 Diesel 37000 Diesel 23000 Diesel 78000 Diesel 47000 Diesel 18000 Diesel 64000 Diesel 50000 Diesel 65000 Diesel 38000 Diesel 70000 Diesel 69000 Diesel 41000 Petrol 16250 Petrol 30676 Diesel 40005 Diesel 38000 Hybrid 45000 Diesel 40005 Diesel 120000 Diesel 51936 Diesel 4000 Diesel 8500 Diesel 119168 Petrol 21800 Petrol 19999 Petrol 1600 Petrol 46221 Diesel 45000 Diesel 40005 Diesel 30676 Diesel 52712 Diesel 59858 Diesel 22 Petrol 28234 Petrol 22803 Diesel 17093 Diesel 8150 Diesel 9344 Diesel null Hybrid 36243 Diesel 33000 Diesel 11300 Petrol 32800 Diesel 37600 Diesel 14000 Petrol 24175 Petrol 27968 Diesel 9255 Diesel 26973 Diesel 12491 Diesel 14999 Diesel 37941 Petrol 11301 Diesel 12629 Diesel 20981 Diesel 23760 Diesel 26216 Diesel 80485 Diesel 18739 Diesel 53453 Diesel 23090 Diesel 26100 Diesel 4285 Diesel 28388 Diesel 16902 Diesel 31227 Diesel 24010 Diesel 15838 Diesel 17657 Hybrid 4977 Diesel 24756 Diesel 41416 Diesel 3814 Petrol 17 Petrol 37080 Diesel 73582 Diesel 23772 Diesel 28808 Diesel 10000 Diesel 26413 Diesel 98000 Diesel 39920 Diesel 10455 Petrol 27644 Diesel 63916 Hybrid 15383 Petrol 21929 Diesel 16900 Diesel 34015 Diesel 2000 Petrol 28000 Petrol 22400 Petrol null Petrol 4500 Petrol 836 Petrol 14462 Diesel 23429 Hybrid 20764 Diesel 31706 Diesel 55000 Diesel 10059 Petrol 10357 Diesel 50374 Diesel 42471 Hybrid 41974 Diesel 20618 Diesel 15536 Petrol 16103 Petrol 14576 Petrol 6198 Diesel 6265 Diesel 21786 Diesel 1558 Petrol 2258 Petrol 26866 Diesel 1887 Petrol 9676 Diesel 7894 Diesel 26179 Diesel 21945 Petrol 2195 Diesel 16291 Petrol 38438 Diesel 25107 Hybrid 20491 Hybrid 52792 Diesel 30035 Petrol 3620 Hybrid 19204 Diesel 10 Petrol 46155 Diesel 49494 Diesel 43014 Diesel 29283 Diesel 38830 Diesel 49761 Diesel 31801 Diesel 22349 Petrol 8520 Diesel 27493 Hybrid 15075 Petrol 5116 Diesel 15009 Petrol 6787 Diesel 32409 Diesel 23690 Diesel 101 Diesel 22485 Hybrid 41921 Petrol 35488 Diesel 41647 Diesel 24923 Diesel 11810 Diesel 2000 Diesel 6668 Diesel 19562 Petrol 29201 Diesel 19864 Diesel 22117 Petrol 43331 Diesel 51929 Diesel 42245 Diesel 42062 Petrol 43883 Diesel 36587 Hybrid 42036 Diesel 36538 Diesel 56000 Diesel 11986 Petrol 38783 Diesel 38051 Diesel 10542 Petrol 3033 Petrol 59884 Diesel 1871 Petrol 32594 Diesel 1474 Petrol 15210 Petrol 49783 Diesel 7175 Diesel null Petrol 35576 Diesel 13888 Diesel 4703 Diesel 8143 Diesel 57066 Diesel 25296 Diesel 29907 Diesel 5659 Petrol 23063 Diesel 60948 Diesel 32769 Diesel 20949 Diesel 27890 Diesel 11954 Diesel 42879 Diesel 11470 Diesel 26205 Diesel 61305 Diesel 10561 Diesel 72939 Hybrid 31125 Diesel 37194 Diesel 2240 Petrol 21023 Diesel 35975 Diesel 55226 Diesel 3686 Diesel 33943 Petrol 9699 Petrol 25059 Diesel 101 Diesel 7006 Petrol 38671 Diesel 19877 Hybrid 7937 Petrol 20027 Diesel 19953 Diesel 18551 Diesel 14288 Diesel 10485 Petrol 101 Diesel 27424 Diesel 2985 Diesel 36051 Diesel 3458 Petrol 6406 Diesel 65000 Diesel 22773 Diesel 37000 Petrol 30487 Diesel 29814 Diesel 28050 Diesel 18882 Diesel 27034 Diesel 40620 Diesel 9057 Diesel 7060 Diesel 4615 Diesel 9250 Petrol 1052 Diesel 25149 Petrol 5947 Petrol 2749 Diesel 7144 Diesel 4843 Diesel 44641 Diesel 26682 Diesel 18865 Diesel 21225 Diesel 52519 Diesel 42000 Diesel 3233 Petrol 28535 Diesel 32629 Diesel 30748 Diesel 17600 

> Same for mileage, distribution is **right skewed** for all fuelType features.

In [0]:
%sql


mileage price 1200 30495 1000 29989 500 37899 5000 30399 4500 29899 null 30999 500 35999 1412 37990 3569 28990 3635 28990 44900 9995 88200 6995 115000 7495 69250 8995 49850 14995 82685 8595 32506 19250 17000 23500 17744 24995 64291 29995 41011 17990 72000 14477 72000 10977 40000 17977 52000 12790 66472 14488 91000 7500 69700 9750 39000 15995 39287 15675 40569 16475 37419 15975 34452 16475 41567 15975 null 16495 37778 17795 44325 17875 45000 14000 75000 16500 36000 21000 64000 15500 106000 7250 67150 13000 43426 11000 42622 13300 34903 16000 35345 12500 23146 14495 16000 24500 44000 23990 30000 17250 22000 14200 19000 22950 90000 10500 32000 15650 90000 14850 37000 22400 23000 19950 78000 9990 47000 11990 18000 20990 64000 14490 50000 10490 65000 15990 38000 17850 70000 16450 69000 11350 41000 12900 16250 26000 30676 22767 40005 12000 38000 18999 45000 14000 40005 12000 120000 4343 51936 16500 4000 27750 8500 28950 119168 3495 21800 59740 19999 44990 1600 54990 46221 15995 45000 14000 40005 12000 30676 22767 52712 12998 59858 17998 22 28998 28234 24998 22803 19498 17093 23498 8150 25498 9344 24998 null 18498 36243 15498 33000 20995 11300 31995 32800 23995 37600 16995 14000 24995 24175 26998 27968 20498 9255 27498 26973 16998 12491 27498 14999 27698 37941 27498 11301 32498 12629 24498 20981 21498 23760 18998 26216 15998 80485 9298 18739 22998 53453 12998 23090 22998 26100 17998 4285 28498 28388 18998 16902 20998 31227 21998 24010 13998 15838 24498 17657 20998 4977 30498 24756 19498 41416 15298 3814 33998 17 32998 37080 16498 73582 15498 23772 15498 28808 18998 10000 24495 26413 18990 98000 10995 39920 15698 10455 22498 27644 19298 63916 16498 15383 19998 21929 20685 16900 20489 34015 13800 2000 31998 28000 29998 22400 24998 null 32998 4500 32998 836 29998 14462 23998 23429 20998 20764 16498 31706 11998 55000 12495 10059 19298 10357 23998 50374 12495 42471 17498 41974 16198 20618 18498 15536 24998 16103 24998 14576 21998 6198 29498 6265 29498 21786 16998 1558 33998 2258 32998 26866 18998 1887 32498 9676 25798 7894 20498 26179 18498 21945 14498 2195 29498 16291 17998 38438 13498 25107 19698 20491 22498 52792 11998 30035 11998 3620 39998 19204 23498 10 57998 46155 16315 49494 13991 43014 16998 29283 16498 38830 16998 49761 13690 31801 14490 22349 17998 8520 29998 27493 18998 15075 18998 5116 28998 15009 24998 6787 27998 32409 16798 23690 19998 101 29950 22485 20790 41921 9490 35488 15690 41647 14990 24923 17290 11810 19490 2000 28990 6668 24990 19562 28990 29201 16298 19864 14698 22117 13998 43331 14698 51929 10498 42245 11998 42062 15698 43883 16998 36587 17998 42036 20998 36538 13691 56000 13491 11986 21998 38783 14500 38051 15990 10542 20490 3033 20890 59884 14490 1871 21990 32594 14750 1474 44990 15210 24690 49783 18270 7175 26625 null 13361 35576 17490 13888 26427 4703 29290 8143 23047 57066 13621 25296 18200 29907 16991 5659 23498 23063 14698 60948 9798 32769 17998 20949 18998 27890 20700 11954 16990 42879 11791 11470 18191 26205 20991 61305 13321 10561 21998 72939 15998 31125 13498 37194 20871 2240 25891 21023 20591 35975 19991 55226 15990 3686 30055 33943 21650 9699 25990 25059 17690 101 30950 7006 24950 38671 14998 19877 19798 7937 24498 20027 19991 19953 21991 18551 22591 14288 18790 10485 25950 101 29950 27424 14990 2985 31950 36051 20990 3458 28990 6406 27000 65000 12590 22773 20885 37000 25885 30487 14965 29814 16595 28050 17940 18882 19940 27034 16498 40620 14500 9057 27791 7060 26499 4615 31990 9250 20999 1052 31999 25149 17495 5947 25495 2749 36995 7144 30999 4843 29999 44641 17880 26682 13660 18865 23920 21225 22960 52519 16590 42000 17990 3233 22270 28535 22680 32629 22440 30748 17820 17600 19290 31360 16390 19598 14450 28192 17250 38087 19690 23523 19250 7716 21980 1647 23990 5572 21980 101 27990 66796 16998 41056 19698 32140 15498 1500 67999 1500 34999 12249 21299 37564 17999 14360 19999 10288 21999 43236 17699 33113 14580 19784 23990 1610 28660 31326 19490 32790 21770 11728 25990 6282 28660 41131 22590 14052 22

> Typical for all used cars listings, price depend strongly by mileage, low mileage the price of car is higher.

In [0]:
%sql


transmission Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Automatic Automatic Automatic Automatic Manual Automatic Automatic Manual Manual Manual Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Manual Automatic Automatic Manual Manual Automatic Manual Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Manual Automatic Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Manual Manual Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Automatic Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Automatic Manual Automatic Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Automatic Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Automatic Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Automatic Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto S

In [0]:
%sql


price engineSize 30495 2.0 29989 1.5 37899 2.0 30399 2.0 29899 2.0 30999 2.0 35999 2.0 37990 3.0 28990 2.0 28990 2.0 9995 1.6 6995 2.1 7495 2.1 8995 2.1 14995 2.1 8595 2.1 19250 2.1 23500 2.1 24995 2.0 29995 4.0 17990 2.0 14477 1.6 10977 2.1 17977 1.6 12790 2.1 14488 2.1 7500 2.1 9750 2.1 15995 1.6 15675 2.1 16475 2.1 15975 2.0 16475 2.1 15975 2.1 16495 2.0 17795 2.1 17875 2.1 14000 2.1 16500 1.6 21000 2.1 15500 2.1 7250 2.1 13000 2.1 11000 1.6 13300 2.1 16000 2.1 12500 2.1 14495 2.0 24500 2.1 23990 6.2 17250 2.1 14200 1.6 22950 2.0 10500 2.1 15650 2.1 14850 2.1 22400 2.1 19950 2.1 9990 2.1 11990 2.1 20990 2.1 14490 2.1 10490 2.1 15990 2.1 17850 2.1 16450 2.1 11350 2.1 12900 1.6 26000 1.5 22767 2.1 12000 2.1 18999 2.0 14000 2.1 12000 2.1 4343 2.1 16500 2.1 27750 2.0 28950 2.0 3495 1.8 59740 4.0 44990 4.0 54990 4.0 15995 2.1 14000 2.1 12000 2.1 22767 2.1 12998 2.1 17998 2.1 28998 1.6 24998 2.0 19498 2.1 23498 2.1 25498 1.6 24998 2.0 18498 2.0 15498 2.1 20995 2.1 31995 3.0 23995 2.1 16995 2.1 24995 2.0 26998 6.2 20498 2.1 27498 2.0 16998 1.6 27498 2.1 27698 2.0 27498 3.0 32498 2.0 24498 2.1 21498 2.1 18998 2.1 15998 2.1 9298 2.1 22998 2.1 12998 2.1 22998 2.1 17998 2.1 28498 2.0 18998 1.6 20998 2.1 21998 2.1 13998 1.6 24498 2.0 20998 2.1 30498 2.0 19498 2.1 15298 2.1 33998 1.5 32998 1.5 16498 2.1 15498 2.1 15498 2.1 18998 2.1 24495 2.0 18990 2.1 10995 2.1 15698 2.1 22498 1.5 19298 2.1 16498 2.0 19998 1.5 20685 2.1 20489 2.1 13800 2.1 31998 2.0 29998 2.0 24998 1.5 32998 2.0 32998 2.0 29998 2.0 23998 2.1 20998 2.0 16498 2.1 11998 2.1 12495 2.1 19298 1.5 23998 2.0 12495 2.1 17498 2.1 16198 2.1 18498 2.1 24998 1.5 24998 1.5 21998 1.5 29498 2.0 29498 2.0 16998 2.1 33998 1.5 32998 1.5 18998 2.1 32498 1.5 25798 2.1 20498 2.1 18498 2.1 14498 2.0 29498 2.0 17998 2.0 13498 2.1 19698 2.0 22498 2.1 11998 1.6 11998 1.6 39998 2.0 23498 2.0 57998 4.0 16315 2.1 13991 2.1 16998 2.1 16498 2.1 16998 2.1 13690 2.1 14490 2.1 17998 2.0 29998 2.0 18998 2.1 18998 2.0 28998 2.0 24998 1.5 27998 2.0 16798 2.1 19998 2.1 29950 2.0 20790 2.1 9490 1.6 15690 2.1 14990 2.1 17290 2.1 19490 2.1 28990 2.0 24990 2.1 28990 6.2 16298 1.6 14698 2.1 13998 2.0 14698 2.1 10498 2.1 11998 2.1 15698 2.0 16998 2.1 17998 2.0 20998 2.1 13691 1.6 13491 2.1 21998 2.0 14500 2.1 15990 2.1 20490 2.0 20890 1.5 14490 2.1 21990 1.5 14750 2.1 44990 3.0 24690 1.5 18270 2.1 26625 2.0 13361 2.0 17490 2.1 26427 2.0 29290 2.0 23047 2.1 13621 1.6 18200 2.1 16991 2.1 23498 2.0 14698 2.1 9798 2.1 17998 2.1 18998 2.1 20700 2.1 16990 2.1 11791 2.1 18191 1.6 20991 2.1 13321 1.6 21998 2.1 15998 2.0 13498 2.1 20871 2.1 25891 2.0 20591 2.1 19991 2.1 15990 2.1 30055 2.0 21650 2.0 25990 2.0 17690 2.1 30950 2.0 24950 1.5 14998 2.1 19798 2.0 24498 1.5 19991 2.1 21991 2.1 22591 2.1 18790 2.1 25950 1.5 29950 2.0 14990 2.1 31950 2.0 20990 2.1 28990 1.5 27000 2.0 12590 2.1 20885 2.1 25885 6.2 14965 2.1 16595 2.1 17940 2.1 19940 2.1 16498 1.6 14500 1.6 27791 2.0 26499 2.0 31990 2.0 20999 1.5 31999 2.0 17495 2.0 25495 1.6 36995 2.0 30999 2.0 29999 2.0 17880 2.1 13660 2.1 23920 2.1 22960 2.1 16590 1.6 17990 2.1 22270 1.6 22680 2.1 22440 2.1 17820 2.1 19290 2.1 16390 2.1 14450 2.0 17250 2.1 19690 2.1 19250 2.1 21980 1.5 23990 1.5 21980 1.5 27990 2.0 16998 2.1 19698 2.1 15498 2.1 67999 4.0 34999 1.5 21299 2.0 17999 2.0 19999 1.5 21999 2.0 17699 2.1 14580 2.1 23990 2.0 28660 2.0 19490 2.1 21770 2.1 25990 2.0 28660 2.0 22590 2.1 22770 2.1 16480 2.1 27698 2.0 16499 2.0 19895 1.5 27595 2.0 25795 2.0 24499 1.5 30495 2.0 29999 2.0 27495 2.0 21795 1.6 9881 1.6 13498 2.0 24450 2.0 17450 2.1 14450 2.1 24990 2.1 16440 2.1 9891 2.1 16998 2.0 16710 2.1 22495 2.1 21565 2.0 21697 2.0 20990 2.1 14770 2.0 13448 2.1 15998 2.1 27498 2.0 19498 2.0 22498 2.1 20105 2.1 14220 2.1 22260 2.1 16990 2.1 41880 3.0 30990 2.0 28880 1.5 28990 2.0 18700 2.1 17440 2.1 17770 2.1 18433 2.1 21440 2.1 22170 2.1 21115 2.1 21497 1.6 28400 2.0 22765 2.1 22880 2.1 22455 1.5 17330 2.1 18795 2.1 15220 2.1 21440 2.1 20200 2.1 19660 1.6 22990 2.1 22770 2.1

In [0]:
%sql


mileage transmission 1200 Automatic 1000 Automatic 500 Automatic 5000 Automatic 4500 Automatic null Automatic 500 Automatic 1412 Automatic 3569 Automatic 3635 Automatic 44900 Automatic 88200 Automatic 115000 Automatic 69250 Automatic 49850 Automatic 82685 Automatic 32506 Semi-Auto 17000 Semi-Auto 17744 Semi-Auto 64291 Semi-Auto 41011 Automatic 72000 Automatic 72000 Automatic 40000 Automatic 52000 Automatic 66472 Automatic 91000 Manual 69700 Automatic 39000 Automatic 39287 Manual 40569 Manual 37419 Manual 34452 Automatic 41567 Automatic null Automatic 37778 Automatic 44325 Automatic 45000 Automatic 75000 Automatic 36000 Automatic 64000 Automatic 106000 Manual 67150 Automatic 43426 Automatic 42622 Manual 34903 Manual 35345 Automatic 23146 Manual 16000 Automatic 44000 Automatic 30000 Automatic 22000 Automatic 19000 Automatic 90000 Automatic 32000 Automatic 90000 Automatic 37000 Automatic 23000 Automatic 78000 Automatic 47000 Automatic 18000 Automatic 64000 Automatic 50000 Automatic 65000 Automatic 38000 Automatic 70000 Automatic 69000 Automatic 41000 Automatic 16250 Automatic 30676 Automatic 40005 Automatic 38000 Automatic 45000 Automatic 40005 Automatic 120000 Automatic 51936 Automatic 4000 Automatic 8500 Automatic 119168 Automatic 21800 Automatic 19999 Automatic 1600 Automatic 46221 Automatic 45000 Automatic 40005 Automatic 30676 Automatic 52712 Semi-Auto 59858 Semi-Auto 22 Manual 28234 Semi-Auto 22803 Semi-Auto 17093 Semi-Auto 8150 Semi-Auto 9344 Semi-Auto null Semi-Auto 36243 Semi-Auto 33000 Semi-Auto 11300 Semi-Auto 32800 Semi-Auto 37600 Automatic 14000 Automatic 24175 Semi-Auto 27968 Semi-Auto 9255 Semi-Auto 26973 Semi-Auto 12491 Semi-Auto 14999 Semi-Auto 37941 Semi-Auto 11301 Semi-Auto 12629 Semi-Auto 20981 Semi-Auto 23760 Semi-Auto 26216 Semi-Auto 80485 Semi-Auto 18739 Semi-Auto 53453 Manual 23090 Semi-Auto 26100 Manual 4285 Semi-Auto 28388 Semi-Auto 16902 Semi-Auto 31227 Semi-Auto 24010 Manual 15838 Semi-Auto 17657 Semi-Auto 4977 Semi-Auto 24756 Semi-Auto 41416 Semi-Auto 3814 Semi-Auto 17 Semi-Auto 37080 Semi-Auto 73582 Semi-Auto 23772 Semi-Auto 28808 Semi-Auto 10000 Semi-Auto 26413 Semi-Auto 98000 Manual 39920 Semi-Auto 10455 Semi-Auto 27644 Semi-Auto 63916 Semi-Auto 15383 Semi-Auto 21929 Semi-Auto 16900 Semi-Auto 34015 Manual 2000 Semi-Auto 28000 Semi-Auto 22400 Semi-Auto null Semi-Auto 4500 Semi-Auto 836 Semi-Auto 14462 Semi-Auto 23429 Semi-Auto 20764 Automatic 31706 Semi-Auto 55000 Automatic 10059 Automatic 10357 Semi-Auto 50374 Semi-Auto 42471 Semi-Auto 41974 Manual 20618 Semi-Auto 15536 Semi-Auto 16103 Semi-Auto 14576 Semi-Auto 6198 Semi-Auto 6265 Semi-Auto 21786 Semi-Auto 1558 Semi-Auto 2258 Semi-Auto 26866 Semi-Auto 1887 Semi-Auto 9676 Semi-Auto 7894 Semi-Auto 26179 Semi-Auto 21945 Manual 2195 Semi-Auto 16291 Semi-Auto 38438 Manual 25107 Automatic 20491 Semi-Auto 52792 Manual 30035 Semi-Auto 3620 Semi-Auto 19204 Semi-Auto 10 Semi-Auto 46155 Semi-Auto 49494 Semi-Auto 43014 Semi-Auto 29283 Semi-Auto 38830 Semi-Auto 49761 Manual 31801 Semi-Auto 22349 Semi-Auto 8520 Semi-Auto 27493 Semi-Auto 15075 Semi-Auto 5116 Semi-Auto 15009 Semi-Auto 6787 Semi-Auto 32409 Semi-Auto 23690 Semi-Auto 101 Semi-Auto 22485 Semi-Auto 41921 Semi-Auto 35488 Semi-Auto 41647 Manual 24923 Semi-Auto 11810 Semi-Auto 2000 Semi-Auto 6668 Automatic 19562 Semi-Auto 29201 Semi-Auto 19864 Manual 22117 Manual 43331 Semi-Auto 51929 Semi-Auto 42245 Semi-Auto 42062 Automatic 43883 Semi-Auto 36587 Semi-Auto 42036 Semi-Auto 36538 Semi-Auto 56000 Manual 11986 Semi-Auto 38783 Semi-Auto 38051 Semi-Auto 10542 Semi-Auto 3033 Semi-Auto 59884 Semi-Auto 1871 Semi-Auto 32594 Semi-Auto 1474 Automatic 15210 Semi-Auto 49783 Semi-Auto 7175 Automatic null Semi-Auto 35576 Semi-Auto 13888 Automatic 4703 Semi-Auto 8143 Semi-Auto 57066 Semi-Auto 25296 Semi-Auto 29907 Semi-Auto 5659 Semi-Auto 23063 Semi-Auto 60948 Semi-Auto 32769 Automatic 20949 Automatic 27890 Semi-Auto 11954 Semi-Auto 42879 Semi-Auto 11470 Semi-Auto 26205 Semi-Auto 61305 Semi-Auto 10561 Semi-Auto 72939 Sem

> Graphs by boxplot shows outliers per each transmission type above 60k miles.

In [0]:
%sql


price fuelType 30495 Diesel 29989 Petrol 37899 Diesel 30399 Diesel 29899 Diesel 30999 Diesel 35999 Diesel 37990 Petrol 28990 Diesel 28990 Diesel 9995 Petrol 6995 Diesel 7495 Diesel 8995 Diesel 14995 Diesel 8595 Diesel 19250 Diesel 23500 Diesel 24995 Diesel 29995 Petrol 17990 Petrol 14477 Diesel 10977 Diesel 17977 Diesel 12790 Diesel 14488 Diesel 7500 Diesel 9750 Diesel 15995 Diesel 15675 Diesel 16475 Diesel 15975 Petrol 16475 Diesel 15975 Hybrid 16495 Hybrid 17795 Diesel 17875 Diesel 14000 Diesel 16500 Diesel 21000 Diesel 15500 Diesel 7250 Diesel 13000 Diesel 11000 Petrol 13300 Diesel 16000 Diesel 12500 Diesel 14495 Petrol 24500 Diesel 23990 Petrol 17250 Diesel 14200 Petrol 22950 Petrol 10500 Diesel 15650 Diesel 14850 Diesel 22400 Diesel 19950 Diesel 9990 Diesel 11990 Diesel 20990 Diesel 14490 Diesel 10490 Diesel 15990 Diesel 17850 Diesel 16450 Diesel 11350 Diesel 12900 Petrol 26000 Petrol 22767 Diesel 12000 Diesel 18999 Hybrid 14000 Diesel 12000 Diesel 4343 Diesel 16500 Diesel 27750 Diesel 28950 Diesel 3495 Petrol 59740 Petrol 44990 Petrol 54990 Petrol 15995 Diesel 14000 Diesel 12000 Diesel 22767 Diesel 12998 Diesel 17998 Diesel 28998 Petrol 24998 Petrol 19498 Diesel 23498 Diesel 25498 Diesel 24998 Diesel 18498 Hybrid 15498 Diesel 20995 Diesel 31995 Petrol 23995 Diesel 16995 Diesel 24995 Petrol 26998 Petrol 20498 Diesel 27498 Diesel 16998 Diesel 27498 Diesel 27698 Diesel 27498 Petrol 32498 Diesel 24498 Diesel 21498 Diesel 18998 Diesel 15998 Diesel 9298 Diesel 22998 Diesel 12998 Diesel 22998 Diesel 17998 Diesel 28498 Diesel 18998 Diesel 20998 Diesel 21998 Diesel 13998 Diesel 24498 Diesel 20998 Hybrid 30498 Diesel 19498 Diesel 15298 Diesel 33998 Petrol 32998 Petrol 16498 Diesel 15498 Diesel 15498 Diesel 18998 Diesel 24495 Diesel 18990 Diesel 10995 Diesel 15698 Diesel 22498 Petrol 19298 Diesel 16498 Hybrid 19998 Petrol 20685 Diesel 20489 Diesel 13800 Diesel 31998 Petrol 29998 Petrol 24998 Petrol 32998 Petrol 32998 Petrol 29998 Petrol 23998 Diesel 20998 Hybrid 16498 Diesel 11998 Diesel 12495 Diesel 19298 Petrol 23998 Diesel 12495 Diesel 17498 Hybrid 16198 Diesel 18498 Diesel 24998 Petrol 24998 Petrol 21998 Petrol 29498 Diesel 29498 Diesel 16998 Diesel 33998 Petrol 32998 Petrol 18998 Diesel 32498 Petrol 25798 Diesel 20498 Diesel 18498 Diesel 14498 Petrol 29498 Diesel 17998 Petrol 13498 Diesel 19698 Hybrid 22498 Hybrid 11998 Diesel 11998 Petrol 39998 Hybrid 23498 Diesel 57998 Petrol 16315 Diesel 13991 Diesel 16998 Diesel 16498 Diesel 16998 Diesel 13690 Diesel 14490 Diesel 17998 Petrol 29998 Diesel 18998 Hybrid 18998 Petrol 28998 Diesel 24998 Petrol 27998 Diesel 16798 Diesel 19998 Diesel 29950 Diesel 20790 Hybrid 9490 Petrol 15690 Diesel 14990 Diesel 17290 Diesel 19490 Diesel 28990 Diesel 24990 Diesel 28990 Petrol 16298 Diesel 14698 Diesel 13998 Petrol 14698 Diesel 10498 Diesel 11998 Diesel 15698 Petrol 16998 Diesel 17998 Hybrid 20998 Diesel 13691 Diesel 13491 Diesel 21998 Petrol 14500 Diesel 15990 Diesel 20490 Petrol 20890 Petrol 14490 Diesel 21990 Petrol 14750 Diesel 44990 Petrol 24690 Petrol 18270 Diesel 26625 Diesel 13361 Petrol 17490 Diesel 26427 Diesel 29290 Diesel 23047 Diesel 13621 Diesel 18200 Diesel 16991 Diesel 23498 Petrol 14698 Diesel 9798 Diesel 17998 Diesel 18998 Diesel 20700 Diesel 16990 Diesel 11791 Diesel 18191 Diesel 20991 Diesel 13321 Diesel 21998 Diesel 15998 Hybrid 13498 Diesel 20871 Diesel 25891 Petrol 20591 Diesel 19991 Diesel 15990 Diesel 30055 Diesel 21650 Petrol 25990 Petrol 17690 Diesel 30950 Diesel 24950 Petrol 14998 Diesel 19798 Hybrid 24498 Petrol 19991 Diesel 21991 Diesel 22591 Diesel 18790 Diesel 25950 Petrol 29950 Diesel 14990 Diesel 31950 Diesel 20990 Diesel 28990 Petrol 27000 Diesel 12590 Diesel 20885 Diesel 25885 Petrol 14965 Diesel 16595 Diesel 17940 Diesel 19940 Diesel 16498 Diesel 14500 Diesel 27791 Diesel 26499 Diesel 31990 Diesel 20999 Petrol 31999 Diesel 17495 Petrol 25495 Petrol 36995 Diesel 30999 Diesel 29999 Diesel 17880 Diesel 13660 Diesel 23920 Diesel 22960 Diesel 16590 Diesel 17990 Di

> As mileage feature above 40k, there are outliers by price.

In [0]:
%sql


model year price transmission mileage fuelType engineSize C Class 2020-01-01 30495 Automatic 1200 Diesel 2.0 C Class 2020-01-01 29989 Automatic 1000 Petrol 1.5 C Class 2020-01-01 37899 Automatic 500 Diesel 2.0 C Class 2019-01-01 30399 Automatic 5000 Diesel 2.0 C Class 2019-01-01 29899 Automatic 4500 Diesel 2.0 C Class 2020-01-01 30999 Automatic null Diesel 2.0 C Class 2020-01-01 35999 Automatic 500 Diesel 2.0 C Class 2019-01-01 37990 Automatic 1412 Petrol 3.0 C Class 2019-01-01 28990 Automatic 3569 Diesel 2.0 C Class 2019-01-01 28990 Automatic 3635 Diesel 2.0 C Class 2013-01-01 9995 Automatic 44900 Petrol 1.6 C Class 2012-01-01 6995 Automatic 88200 Diesel 2.1 C Class 2012-01-01 7495 Automatic 115000 Diesel 2.1 C Class 2011-01-01 8995 Automatic 69250 Diesel 2.1 C Class 2015-01-01 14995 Automatic 49850 Diesel 2.1 C Class 2013-01-01 8595 Automatic 82685 Diesel 2.1 C Class 2016-01-01 19250 Semi-Auto 32506 Diesel 2.1 C Class 2017-01-01 23500 Semi-Auto 17000 Diesel 2.1 C Class 2018-01-01 24995 Semi-Auto 17744 Diesel 2.0 C Class 2016-01-01 29995 Semi-Auto 64291 Petrol 4.0 C Class 2014-01-01 17990 Automatic 41011 Petrol 2.0 C Class null 14477 Automatic 72000 Diesel 1.6 C Class 2013-01-01 10977 Automatic 72000 Diesel 2.1 C Class 2016-01-01 17977 Automatic 40000 Diesel 1.6 C Class 2015-01-01 12790 Automatic 52000 Diesel 2.1 C Class 2017-01-01 14488 Automatic 66472 Diesel 2.1 C Class 2011-01-01 7500 Manual 91000 Diesel 2.1 C Class 2013-01-01 9750 Automatic 69700 Diesel 2.1 C Class 2015-01-01 15995 Automatic 39000 Diesel 1.6 C Class 2016-01-01 15675 Manual 39287 Diesel 2.1 C Class 2016-01-01 16475 Manual 40569 Diesel 2.1 C Class 2016-01-01 15975 Manual 37419 Petrol 2.0 C Class 2016-01-01 16475 Automatic 34452 Diesel 2.1 C Class 2015-01-01 15975 Automatic 41567 Hybrid 2.1 C Class 2015-01-01 16495 Automatic null Hybrid 2.0 C Class 2017-01-01 17795 Automatic 37778 Diesel 2.1 C Class 2017-01-01 17875 Automatic 44325 Diesel 2.1 C Class 2016-01-01 14000 Automatic 45000 Diesel 2.1 C Class 2017-01-01 16500 Automatic 75000 Diesel 1.6 C Class 2016-01-01 21000 Automatic 36000 Diesel 2.1 C Class 2015-01-01 15500 Automatic 64000 Diesel 2.1 C Class 2013-01-01 7250 Manual 106000 Diesel 2.1 C Class 2015-01-01 13000 Automatic 67150 Diesel 2.1 C Class 2012-01-01 11000 Automatic 43426 Petrol 1.6 C Class 2016-01-01 13300 Manual 42622 Diesel 2.1 C Class 2016-01-01 16000 Manual 34903 Diesel 2.1 C Class 2014-01-01 12500 Automatic 35345 Diesel 2.1 C Class 2015-01-01 14495 Manual 23146 Petrol 2.0 C Class 2017-01-01 24500 Automatic 16000 Diesel 2.1 C Class 2013-01-01 23990 Automatic 44000 Petrol 6.2 C Class 2015-01-01 17250 Automatic 30000 Diesel 2.1 C Class 2015-01-01 14200 Automatic 22000 Petrol 1.6 C Class 2017-01-01 22950 Automatic 19000 Petrol 2.0 C Class 2013-01-01 10500 Automatic 90000 Diesel 2.1 C Class 2014-01-01 15650 Automatic 32000 Diesel 2.1 C Class 2015-01-01 14850 Automatic 90000 Diesel 2.1 C Class 2017-01-01 22400 Automatic 37000 Diesel 2.1 C Class 2017-01-01 19950 Automatic 23000 Diesel 2.1 C Class 2014-01-01 9990 Automatic 78000 Diesel 2.1 C Class 2015-01-01 11990 Automatic 47000 Diesel 2.1 C Class 2017-01-01 20990 Automatic 18000 Diesel 2.1 C Class 2015-01-01 14490 Automatic 64000 Diesel 2.1 C Class 2013-01-01 10490 Automatic 50000 Diesel 2.1 C Class 2016-01-01 15990 Automatic 65000 Diesel 2.1 C Class 2016-01-01 17850 Automatic 38000 Diesel 2.1 C Class 2016-01-01 16450 Automatic 70000 Diesel 2.1 C Class 2013-01-01 11350 Automatic 69000 Diesel 2.1 C Class 2014-01-01 12900 Automatic 41000 Petrol 1.6 C Class 2018-01-01 26000 Automatic 16250 Petrol 1.5 C Class 2017-01-01 22767 Automatic 30676 Diesel 2.1 C Class 2015-01-01 12000 Automatic 40005 Diesel 2.1 C Class 2017-01-01 18999 Automatic 38000 Hybrid 2.0 C Class 2016-01-01 14000 Automatic 45000 Diesel 2.1 C Class 2015-01-01 12000 Automatic 40005 Diesel 2.1 C Class 2009-01-01 4343 Automatic 120000 Diesel 2.1 C Class 2016-01-01 16500 Automatic 51936 Diesel 2.1 C Class 2019-01-01 27750 Automatic 4000 

In [0]:
%sql


model year price transmission mileage fuelType engineSize C Class 2020-01-01 30495 Automatic 1200 Diesel 2.0 C Class 2020-01-01 29989 Automatic 1000 Petrol 1.5 C Class 2020-01-01 37899 Automatic 500 Diesel 2.0 C Class 2019-01-01 30399 Automatic 5000 Diesel 2.0 C Class 2019-01-01 29899 Automatic 4500 Diesel 2.0 C Class 2020-01-01 30999 Automatic null Diesel 2.0 C Class 2020-01-01 35999 Automatic 500 Diesel 2.0 C Class 2019-01-01 37990 Automatic 1412 Petrol 3.0 C Class 2019-01-01 28990 Automatic 3569 Diesel 2.0 C Class 2019-01-01 28990 Automatic 3635 Diesel 2.0 C Class 2013-01-01 9995 Automatic 44900 Petrol 1.6 C Class 2012-01-01 6995 Automatic 88200 Diesel 2.1 C Class 2012-01-01 7495 Automatic 115000 Diesel 2.1 C Class 2011-01-01 8995 Automatic 69250 Diesel 2.1 C Class 2015-01-01 14995 Automatic 49850 Diesel 2.1 C Class 2013-01-01 8595 Automatic 82685 Diesel 2.1 C Class 2016-01-01 19250 Semi-Auto 32506 Diesel 2.1 C Class 2017-01-01 23500 Semi-Auto 17000 Diesel 2.1 C Class 2018-01-01 24995 Semi-Auto 17744 Diesel 2.0 C Class 2016-01-01 29995 Semi-Auto 64291 Petrol 4.0 C Class 2014-01-01 17990 Automatic 41011 Petrol 2.0 C Class null 14477 Automatic 72000 Diesel 1.6 C Class 2013-01-01 10977 Automatic 72000 Diesel 2.1 C Class 2016-01-01 17977 Automatic 40000 Diesel 1.6 C Class 2015-01-01 12790 Automatic 52000 Diesel 2.1 C Class 2017-01-01 14488 Automatic 66472 Diesel 2.1 C Class 2011-01-01 7500 Manual 91000 Diesel 2.1 C Class 2013-01-01 9750 Automatic 69700 Diesel 2.1 C Class 2015-01-01 15995 Automatic 39000 Diesel 1.6 C Class 2016-01-01 15675 Manual 39287 Diesel 2.1 C Class 2016-01-01 16475 Manual 40569 Diesel 2.1 C Class 2016-01-01 15975 Manual 37419 Petrol 2.0 C Class 2016-01-01 16475 Automatic 34452 Diesel 2.1 C Class 2015-01-01 15975 Automatic 41567 Hybrid 2.1 C Class 2015-01-01 16495 Automatic null Hybrid 2.0 C Class 2017-01-01 17795 Automatic 37778 Diesel 2.1 C Class 2017-01-01 17875 Automatic 44325 Diesel 2.1 C Class 2016-01-01 14000 Automatic 45000 Diesel 2.1 C Class 2017-01-01 16500 Automatic 75000 Diesel 1.6 C Class 2016-01-01 21000 Automatic 36000 Diesel 2.1 C Class 2015-01-01 15500 Automatic 64000 Diesel 2.1 C Class 2013-01-01 7250 Manual 106000 Diesel 2.1 C Class 2015-01-01 13000 Automatic 67150 Diesel 2.1 C Class 2012-01-01 11000 Automatic 43426 Petrol 1.6 C Class 2016-01-01 13300 Manual 42622 Diesel 2.1 C Class 2016-01-01 16000 Manual 34903 Diesel 2.1 C Class 2014-01-01 12500 Automatic 35345 Diesel 2.1 C Class 2015-01-01 14495 Manual 23146 Petrol 2.0 C Class 2017-01-01 24500 Automatic 16000 Diesel 2.1 C Class 2013-01-01 23990 Automatic 44000 Petrol 6.2 C Class 2015-01-01 17250 Automatic 30000 Diesel 2.1 C Class 2015-01-01 14200 Automatic 22000 Petrol 1.6 C Class 2017-01-01 22950 Automatic 19000 Petrol 2.0 C Class 2013-01-01 10500 Automatic 90000 Diesel 2.1 C Class 2014-01-01 15650 Automatic 32000 Diesel 2.1 C Class 2015-01-01 14850 Automatic 90000 Diesel 2.1 C Class 2017-01-01 22400 Automatic 37000 Diesel 2.1 C Class 2017-01-01 19950 Automatic 23000 Diesel 2.1 C Class 2014-01-01 9990 Automatic 78000 Diesel 2.1 C Class 2015-01-01 11990 Automatic 47000 Diesel 2.1 C Class 2017-01-01 20990 Automatic 18000 Diesel 2.1 C Class 2015-01-01 14490 Automatic 64000 Diesel 2.1 C Class 2013-01-01 10490 Automatic 50000 Diesel 2.1 C Class 2016-01-01 15990 Automatic 65000 Diesel 2.1 C Class 2016-01-01 17850 Automatic 38000 Diesel 2.1 C Class 2016-01-01 16450 Automatic 70000 Diesel 2.1 C Class 2013-01-01 11350 Automatic 69000 Diesel 2.1 C Class 2014-01-01 12900 Automatic 41000 Petrol 1.6 C Class 2018-01-01 26000 Automatic 16250 Petrol 1.5 C Class 2017-01-01 22767 Automatic 30676 Diesel 2.1 C Class 2015-01-01 12000 Automatic 40005 Diesel 2.1 C Class 2017-01-01 18999 Automatic 38000 Hybrid 2.0 C Class 2016-01-01 14000 Automatic 45000 Diesel 2.1 C Class 2015-01-01 12000 Automatic 40005 Diesel 2.1 C Class 2009-01-01 4343 Automatic 120000 Diesel 2.1 C Class 2016-01-01 16500 Automatic 51936 Diesel 2.1 C Class 2019-01-01 27750 Automatic 4000 

> Car used price depending by mileage and age, when both are low, price is higher.

In [0]:
%sql


date_format(CAST(current_date() AS TIMESTAMP), y) 2021

In [0]:
%sql


price carAge 30495 1 29989 1 37899 1 30399 2 29899 2 30999 1 35999 1 37990 2 28990 2 28990 2 9995 8 6995 9 7495 9 8995 10 14995 6 8595 8 19250 5 23500 4 24995 3 29995 5 17990 7 14477 null 10977 8 17977 5 12790 6 14488 4 7500 10 9750 8 15995 6 15675 5 16475 5 15975 5 16475 5 15975 6 16495 6 17795 4 17875 4 14000 5 16500 4 21000 5 15500 6 7250 8 13000 6 11000 9 13300 5 16000 5 12500 7 14495 6 24500 4 23990 8 17250 6 14200 6 22950 4 10500 8 15650 7 14850 6 22400 4 19950 4 9990 7 11990 6 20990 4 14490 6 10490 8 15990 5 17850 5 16450 5 11350 8 12900 7 26000 3 22767 4 12000 6 18999 4 14000 5 12000 6 4343 12 16500 5 27750 2 28950 2 3495 14 59740 4 44990 2 54990 2 15995 6 14000 5 12000 6 22767 4 12998 7 17998 4 28998 2 24998 4 19498 5 23498 3 25498 2 24998 2 18498 5 15498 7 20995 4 31995 3 23995 4 16995 7 24995 4 26998 7 20498 5 27498 3 16998 5 27498 3 27698 2 27498 5 32498 2 24498 2 21498 4 18998 4 15998 7 9298 8 22998 3 12998 5 22998 5 17998 5 28498 2 18998 5 20998 4 21998 5 13998 6 24498 2 20998 4 30498 2 19498 5 15298 7 33998 2 32998 2 16498 5 15498 5 15498 7 18998 4 24495 2 18990 5 10995 7 15698 7 22498 2 19298 5 16498 4 19998 3 20685 3 20489 3 13800 4 31998 2 29998 2 24998 2 32998 2 32998 2 29998 2 23998 4 20998 5 16498 4 11998 8 12495 6 19298 2 23998 2 12495 6 17498 6 16198 5 18498 4 24998 2 24998 2 21998 2 29498 3 29498 3 16998 5 33998 2 32998 2 18998 4 32498 2 25798 4 20498 3 18498 4 14498 5 29498 2 17998 4 13498 6 19698 5 22498 4 11998 6 11998 null 39998 1 23498 2 57998 2 16315 6 13991 5 16998 5 16498 4 16998 5 13690 5 14490 7 17998 4 29998 2 18998 4 18998 3 28998 2 24998 2 27998 3 16798 6 19998 4 29950 1 20790 4 9490 8 15690 4 14990 4 17290 6 19490 3 28990 2 24990 4 28990 7 16298 5 14698 5 13998 5 14698 6 10498 7 11998 6 15698 5 16998 6 17998 5 20998 4 13691 5 13491 7 21998 4 14500 6 15990 4 20490 5 20890 3 14490 7 21990 3 14750 6 44990 2 24690 2 18270 4 26625 3 13361 6 17490 4 26427 2 29290 2 23047 4 13621 6 18200 4 16991 4 23498 3 14698 6 9798 8 17998 5 18998 4 20700 5 16990 6 11791 7 18191 6 20991 5 13321 4 21998 5 15998 4 13498 6 20871 4 25891 3 20591 4 19991 5 15990 6 30055 2 21650 4 25990 3 17690 4 30950 2 24950 2 14998 6 19798 4 24498 2 19991 3 21991 4 22591 3 18790 2 25950 2 29950 1 14990 4 31950 1 20990 5 28990 1 27000 2 12590 6 20885 3 25885 7 14965 5 16595 4 17940 3 19940 3 16498 5 14500 5 27791 2 26499 2 31990 2 20999 3 31999 2 17495 4 25495 2 36995 2 30999 2 29999 2 17880 5 13660 5 23920 5 22960 4 16590 4 17990 4 22270 2 22680 5 22440 4 17820 5 19290 5 16390 5 14450 5 17250 5 19690 5 19250 4 21980 2 23990 2 21980 2 27990 2 16998 6 19698 5 15498 6 67999 1 34999 1 21299 5 17999 4 19999 2 21999 4 17699 5 14580 5 23990 4 28660 2 19490 4 21770 4 25990 2 28660 2 22590 4 22770 4 16480 6 27698 3 16499 4 19895 2 27595 2 25795 2 24499 2 30495 2 29999 2 27495 3 21795 2 9881 8 13498 6 24450 2 17450 6 14450 5 24990 4 16440 6 9891 7 16998 6 16710 6 22495 4 21565 4 21697 4 20990 5 14770 5 13448 6 15998 7 27498 3 19498 5 22498 4 20105 5 14220 5 22260 4 16990 7 41880 2 30990 2 28880 2 28990 2 18700 5 17440 4 17770 5 18433 4 21440 5 22170 4 21115 4 21497 2 28400 2 22765 5 22880 3 22455 2 17330 4 18795 4 15220 5 21440 4 20200 4 19660 3 22990 4 22770 3 19775 3 24660 4 33495 4 26890 2 26885 2 34790 2 18495 4 15699 5 18499 4 24799 3 18995 4 15280 7 14795 7 19840 3 13791 7 24360 4 16990 5 17990 5 14990 7 15990 5 16998 4 21495 5 20995 5 20495 3 24736 2 21595 4 21295 4 19995 5 23495 4 15490 5 17000 4 16402 4 20682 2 29232 3 14698 6 14288 7 18265 3 15971 7 19465 4 16210 4 16990 5 20990 4 17699 5 18495 4 28995 2 14495 5 13950 7 12750 7 19495 5 18995 5 19950 3 19498 4 16360 7 16767 4 10190 7 19298 3 15998 6 15998 5 24698 4 14398 5 15998 7 19950 3 11149 7 17990 3 16490 4 18450 4 11850 6 16350 4 10477 6 18998 3 15870 5 16700 5 19502 4 28332 2 17992 5 49280 2 28810 2 30000 2 17901 6 19971 4 26281 3 17791 4 16998 5 14498 7 16998 5 16990 6 19798 5 16998 5 16750 4 12880 6 18660 4 17990 5 25882 3 29046 2 26256 2 20498 4 11298 8 16500 4 16990 7 2

> Confirm when lower the car age is, the price is higher.

## 6. Data Wrangling

### 6.1 Missing Values

Many real-world datasets may contain missing values for various reasons. They are often encoded as NaNs, blanks or any other placeholders. Training a model with a dataset that has a lot of missing values can drastically impact the machine learning model’s quality. Some algorithms such as scikit-learn estimators assume that all values are numerical and have and hold meaningful value.

One way to handle this problem is to get rid of the observations that have missing data. However, you will risk losing data points with valuable information. A better strategy would be to impute the missing values. In other words, we need to infer those missing values from the existing part of the data. 

we will focus on 6 popular ways for data imputation for cross-sectional datasets:

* Do Nothing
* Imputation Using (Mean/Median) Values
* Imputation Using (Most Frequent) or (Zero/Constant) Values
* Imputation Using k-NN
* Imputation Using Multivariate Imputation by Chained Equation (MICE)
* Imputation Using Deep Learning (Datawig)

We can check all of them and evaluate model combining each step described before.

#### a. Do Nothing

That’s an easy one. You just let the algorithm handle the missing data. Some algorithms can factor in the missing values and learn the best imputation values for the missing data based on the training loss reduction (ie. XGBoost). Some others have the option to just ignore them (ie. LightGBM — use_missing=false). However, other algorithms will panic and throw an error complaining about the missing values (ie. Scikit learn — LinearRegression). In that case, you will need to handle the missing data and clean it before feeding it to the algorithm.

#### b. Imputation Using (Mean/Median) Values:

This works by calculating the mean/median of the non-missing values in a column and then replacing the missing values within each column separately and independently from the others. It can only be used with numeric data.

*Pros*:
Easy and fast.
Works well with small numerical datasets.

*Cons*:
Doesn’t factor the correlations between features. It only works on the column level.
Will give poor results on encoded categorical features (do NOT use it on categorical features).
Not very accurate.
Doesn’t account for the uncertainty in the imputations

#### c. Imputation Using (Most Frequent) or (Zero/Constant) Values

Most Frequent is another statistical strategy to impute missing values and YES!! It works with categorical features (strings or numerical representations) by replacing missing data with the most frequent values within each column.

*Pros*:
Works well with categorical features.

*Cons*:
It also doesn’t factor the correlations between features.
It can introduce bias in the data.

#### d. Imputation Using k-NN:

The k nearest neighbours is an algorithm that is used for simple classification. The algorithm uses ‘feature similarity’ to predict the values of any new data points. This means that the new point is assigned a value based on how closely it resembles the points in the training set. This can be very useful in making predictions about the missing values by finding the k’s closest neighbours to the observation with missing data and then imputing them based on the non-missing values in the neighbourhood. Let’s see some example code using Impyute library which provides a simple and easy way to use KNN for imputation:

*Pros*:
Can be much more accurate than the mean, median or most frequent imputation methods (It depends on the dataset).

*Cons*:
Computationally expensive. KNN works by storing the whole training dataset in memory.
K-NN is quite sensitive to outliers in the data (unlike SVM)

#### e. Imputation Using Multivariate Imputation by Chained Equation (MICE)

This type of imputation works by filling the missing data multiple times. Multiple Imputations (MIs) are much better than a single imputation as it measures the uncertainty of the missing values in a better way. The chained equations approach is also very flexible and can handle different variables of different data types (ie., continuous or binary) as well as complexities such as bounds or survey skip patterns. For more information on the algorithm mechanics, you can refer to the Research Paper

#### f. Imputation Using Deep Learning (Datawig):

This method works very well with categorical and non-numerical features. It is a library that learns Machine Learning models using Deep Neural Networks to impute missing values in a dataframe. It also supports both CPU and GPU for training.

*Pros*:
Quite accurate compared to other methods.
It has some functions that can handle categorical data (Feature Encoder).
It supports CPUs and GPUs.

*Cons*:
Single Column imputation.
Can be quite slow with large datasets.
You have to specify the columns that contain information about the target column that will be imputed.


#### Other Imputation Methods:

##### Stochastic regression imputation:
It is quite similar to regression imputation which tries to predict the missing values by regressing it from other related variables in the same dataset plus some random residual value.

##### Extrapolation and Interpolation:
It tries to estimate values from other observations within the range of a discrete set of known data points.

##### Hot-Deck imputation:
Works by randomly choosing the missing value from a set of related and similar variables.

In [0]:
%python
# Check null and NA values for both dataset by frequency first and absolute values after
# Create a function to detect and shows missing values


In [0]:
# Check null values
import pandas as pd


Your selected dataframe has 7 columns and 3899 Rows.
There are 2 columns that have missing values.
Out[10]:

Zero Values 
 Missing Values 
 % of Total Values 
 Total Zero Missing Values 
 % Total Zero Missing Values 
 Data Type 
 
 
 
 
 mileage 
 0 
 7 
 0.2 
 7 
 0.2 
 float64 
 
 
 year 
 0 
 5 
 0.1 
 5 
 0.1 
 object

> It seems dataset has only 12 missing values, 7 in `mileage` and 5 in `year` feature. We analyze them to take a decision to impute or remove them.

In [0]:
%sql
/*# Check year null values on cross table */


year total_values 2019 1589 2017 665 2016 544 2018 361 2015 321 2014 136 2020 125 2013 63 2012 26 2011 13 2010 10 2008 8 2007 7 2009 7 null 5 2005 5 2002 4 2004 3 2006 3 2003 1 1991 1 1998 1 1995 1

In [0]:
%sql
/* # Check null values and all variables */


model year price transmission mileage fuelType engineSize C Class null 14477 Automatic 72000 Diesel 1.6 C Class null 11998 Semi-Auto 30035 Petrol 1.6 C Class null 37995 Semi-Auto 5000 Petrol 1.5 C Class null 28180 Automatic 11487 Diesel 2.0 C Class null 18500 Automatic 28248 Diesel 2.1

In [0]:
%sql
/* first step to understanding unique values */


distinct_values_mileage 22

> There are only `22` unique years in the dataset. The idea to create cluster with not null values, understand distribution and impute within cluster associated by other features (price, mileage, fuelType, engineSize). To re-create the same scenario, we take the first null year value, and compare with other not null value and analyze them.

> Check first null value
```
C Class
null
14477
Automatic
72000
Diesel
1.6
```

In [0]:
%sql
/* Checking the first null value above, we select only price between 12k and 15k, automatic transmission, diesel fuelType, 1.6 engineSize and mileage between 70k and 74k */


model year price transmission mileage fuelType engineSize

> Check with other wide range, maybe we haven't same characteristics like null value.

In [0]:
%sql


model year price transmission mileage fuelType engineSize C Class 2017-01-01 16500 Automatic 75000 Diesel 1.6

> And we have only one value with similar values, so in conclusion firt null value will be impute to `2017` ?

> With same approach, we analyze same procedure with the second null value, mileage.

In [0]:
%sql
/*# Check mileage null values cross table */


mileage total_values 1000 82 2000 21 16 20 3000 17 100 17 40005 17 45000 16 9 15 6000 14 20000 11 30676 11 54000 10 5000 10 4999 10 500 9 15000 9 3999 8 22000 8 4000 8 29000 8

In [0]:
%sql
/* check how many distinct values exists */


distinct_values_mileage 3199

> There are only `3199` unique values for mileage feature. Of course, isn't discrete variable as year, but **continuous variable**, and so wide range to cluster or predict and impute null values.

In [0]:
%sql
/* # Check null values and all attributes for mileage feature */


model year price transmission mileage fuelType engineSize C Class 2020-01-01 30999 Automatic null Diesel 2.0 C Class 2015-01-01 16495 Automatic null Hybrid 2.0 C Class 2016-01-01 18498 Semi-Auto null Hybrid 2.0 C Class 2019-01-01 32998 Semi-Auto null Petrol 2.0 C Class 2015-01-01 13361 Semi-Auto null Petrol 2.0 C Class 2015-01-01 18299 Automatic null Diesel 2.1 C Class 2019-01-01 24995 Automatic null Petrol 1.5

> For instance, if we impute null values with median or mean, or we want to clusterize in groups to impute them with other techniques described above, regression for example, we could check per each observation distribution within group. Manually check and investigate the variables.

In [0]:
%sql
/* For the first null value above, we select only price between 30k and 32k, automatic transmission, diesel fuelType, 2.0 engineSize and year 2020 */


model year price transmission mileage fuelType engineSize C Class 2020-01-01 30495 Automatic 1200 Diesel 2.0 C Class 2020-01-01 31950 Automatic 2985 Diesel 2.0 C Class 2020-01-01 30990 Automatic 6000 Diesel 2.0 C Class 2020-01-01 31690 Automatic 3000 Diesel 2.0 C Class 2020-01-01 31980 Automatic 500 Diesel 2.0 C Class 2020-01-01 31590 Automatic 1452 Diesel 2.0 C Class 2020-01-01 30395 Automatic 2500 Diesel 2.0 C Class 2020-01-01 30480 Automatic 500 Diesel 2.0 C Class 2020-01-01 31900 Automatic 5000 Diesel 2.0

In [0]:
%sql
/* Visualize the price and mileage mean */


model year price transmission mileage fuelType engineSize C Class 2020-01-01 30495 Automatic 1200 Diesel 2.0 C Class 2020-01-01 31950 Automatic 2985 Diesel 2.0 C Class 2020-01-01 30990 Automatic 6000 Diesel 2.0 C Class 2020-01-01 31690 Automatic 3000 Diesel 2.0 C Class 2020-01-01 31980 Automatic 500 Diesel 2.0 C Class 2020-01-01 31590 Automatic 1452 Diesel 2.0 C Class 2020-01-01 30395 Automatic 2500 Diesel 2.0 C Class 2020-01-01 30480 Automatic 500 Diesel 2.0 C Class 2020-01-01 31900 Automatic 5000 Diesel 2.0

> We catch anormal distribution by mean of price, because probably depend the car features, and for instance, car model. In this dataset we have only feature represnt C-Class, not specified which kind C-Class model is.

***
Conclusion
> After checking year and mileage null values like steps described above to understanding distribution and take decision to remove them.

In [0]:
# Create a backup dataset


In [0]:
# Check df_bak structure


Out[12]: 3899

In [0]:
# Drop Null values


In [0]:
# Check after null values removed


Out[14]: 3887

## 7. Feature Engineering

`Feature engineering` is about creating new input features from your existing ones.

In general, you can think of data cleaning as a process of subtraction and feature engineering as a process of addition.

This is often one of the most valuable tasks a data science area can do to improve model performance, for 3 big reasons:

- we can isolate and highlight key information, which helps your algorithms "focus" on what’s important.
- we can bring in your own domain expertise.
- most importantly, once you understand the "vocabulary" of feature engineering, you can bring in other people’s domain expertise!

Main tasks are :

* 1. combine sparse classes
* 2. add dummy variables
* 3. remove unused features

In [0]:
# df.show()

+-------+----------+-----+------------+-------+--------+----------+
 model| year|price|transmission|mileage|fuelType|engineSize|
+-------+----------+-----+------------+-------+--------+----------+
C Class|2020-01-01|30495| Automatic| 1200| Diesel| 2.0|
C Class|2020-01-01|29989| Automatic| 1000| Petrol| 1.5|
C Class|2020-01-01|37899| Automatic| 500| Diesel| 2.0|
C Class|2019-01-01|30399| Automatic| 5000| Diesel| 2.0|
C Class|2019-01-01|29899| Automatic| 4500| Diesel| 2.0|
C Class|2020-01-01|35999| Automatic| 500| Diesel| 2.0|
C Class|2019-01-01|37990| Automatic| 1412| Petrol| 3.0|
C Class|2019-01-01|28990| Automatic| 3569| Diesel| 2.0|
C Class|2019-01-01|28990| Automatic| 3635| Diesel| 2.0|
C Class|2013-01-01| 9995| Automatic| 44900| Petrol| 1.6|
C Class|2012-01-01| 6995| Automatic| 88200| Diesel| 2.1|
C Class|2012-01-01| 7495| Automatic| 115000| Diesel| 2.1|
C Class|2011-01-01| 8995| Automatic| 69250| Diesel| 2.1|
C Class|2015-01-01|14995| Automatic| 49850| Diesel| 2.1|
C Class|2013-01-01| 8595| Automatic| 82685| Diesel| 2.1|
C Class|2016-01-01|19250| Semi-Auto| 32506| Diesel| 2.1|
C Class|2017-01-01|23500| Semi-Auto| 17000| Diesel| 2.1|
C Class|2018-01-01|24995| Semi-Auto| 17744| Diesel| 2.0|
C Class|2016-01-01|29995| Semi-Auto| 64291| Petrol| 4.0|
C Class|2014-01-01|17990| Automatic| 41011| Petrol| 2.0|
+-------+----------+-----+------------+-------+--------+----------+
only showing top 20 rows

In [0]:
# Create new feature, carAge calculate the age of car.
from pyspark.sql.functions import year, datediff, date_format, current_date

In [0]:
# df.show(5)

+-------+----------+-----+------------+-------+--------+----------+------+
 model| year|price|transmission|mileage|fuelType|engineSize|carAge|
+-------+----------+-----+------------+-------+--------+----------+------+
C Class|2020-01-01|30495| Automatic| 1200| Diesel| 2.0| 1.0|
C Class|2020-01-01|29989| Automatic| 1000| Petrol| 1.5| 1.0|
C Class|2020-01-01|37899| Automatic| 500| Diesel| 2.0| 1.0|
C Class|2019-01-01|30399| Automatic| 5000| Diesel| 2.0| 2.0|
C Class|2019-01-01|29899| Automatic| 4500| Diesel| 2.0| 2.0|
+-------+----------+-----+------------+-------+--------+----------+------+
only showing top 5 rows

In [0]:
# Remove null values


Out[18]: DataFrame[model: string, year: date, price: int, transmission: string, mileage: int, fuelType: string, engineSize: string, carAge: double]

In [0]:
# Remove columns year and model because aren't important features


Out[20]: 3887

In [0]:
%sql
/* Show database and tables */


databaseName default

In [0]:
%sql

database tableName isTemporary cclass_csv true

In [0]:
# Remove temporary parquet location

Out[27]: True

In [0]:
# With this registered as a temp view, it will only be available to this particular notebook. If you'd like other users to be able to query this table, you can also create a table from the DataFrame.
# Once saved, this table will persist across cluster restarts as well as allow various users across different notebooks to query this data.
# To do so, choose your table name and uncomment the bottom line.



In [0]:
%sql


price transmission mileage fuelType engineSize carAge 30495 Automatic 1200 Diesel 2.0 1.0 29989 Automatic 1000 Petrol 1.5 1.0 37899 Automatic 500 Diesel 2.0 1.0

## 8. Batch Inference - Delta Table

There are many scenarios where you might want to evaluate a model on a corpus of new data. For example, you may have a fresh batch of data, or may need to compare the performance of two models on the same corpus of data.

The following code evaluates the model on data stored in a Delta table, using Spark to run the computation in parallel.

In [0]:
# To simulate a new corpus of data, save the existing DF data to a Delta table. 
# In the real world, this would be a new batch of data.
#spark_df = spark.createDataFrame(df)

# Replace <username> with your username before running this cell.


# Delete the contents of this path in case this cell has already been run


> Load the model into a Spark UDF, so it can be applied to the Delta table.

In [0]:
# Read the "new data" from Delta


In [0]:
# show results


price transmission mileage fuelType engineSize carAge 30495 Automatic 1200 Diesel 2.0 1.0 29989 Automatic 1000 Petrol 1.5 1.0 37899 Automatic 500 Diesel 2.0 1.0 30399 Automatic 5000 Diesel 2.0 2.0 29899 Automatic 4500 Diesel 2.0 2.0 35999 Automatic 500 Diesel 2.0 1.0 37990 Automatic 1412 Petrol 3.0 2.0 28990 Automatic 3569 Diesel 2.0 2.0 28990 Automatic 3635 Diesel 2.0 2.0 9995 Automatic 44900 Petrol 1.6 8.0 6995 Automatic 88200 Diesel 2.1 9.0 7495 Automatic 115000 Diesel 2.1 9.0 8995 Automatic 69250 Diesel 2.1 10.0 14995 Automatic 49850 Diesel 2.1 6.0 8595 Automatic 82685 Diesel 2.1 8.0 19250 Semi-Auto 32506 Diesel 2.1 5.0 23500 Semi-Auto 17000 Diesel 2.1 4.0 24995 Semi-Auto 17744 Diesel 2.0 3.0 29995 Semi-Auto 64291 Petrol 4.0 5.0 17990 Automatic 41011 Petrol 2.0 7.0 10977 Automatic 72000 Diesel 2.1 8.0 17977 Automatic 40000 Diesel 1.6 5.0 12790 Automatic 52000 Diesel 2.1 6.0 14488 Automatic 66472 Diesel 2.1 4.0 7500 Manual 91000 Diesel 2.1 10.0 9750 Automatic 69700 Diesel 2.1 8.0 15995 Automatic 39000 Diesel 1.6 6.0 15675 Manual 39287 Diesel 2.1 5.0 16475 Manual 40569 Diesel 2.1 5.0 15975 Manual 37419 Petrol 2.0 5.0 16475 Automatic 34452 Diesel 2.1 5.0 15975 Automatic 41567 Hybrid 2.1 6.0 17795 Automatic 37778 Diesel 2.1 4.0 17875 Automatic 44325 Diesel 2.1 4.0 14000 Automatic 45000 Diesel 2.1 5.0 16500 Automatic 75000 Diesel 1.6 4.0 21000 Automatic 36000 Diesel 2.1 5.0 15500 Automatic 64000 Diesel 2.1 6.0 7250 Manual 106000 Diesel 2.1 8.0 13000 Automatic 67150 Diesel 2.1 6.0 11000 Automatic 43426 Petrol 1.6 9.0 13300 Manual 42622 Diesel 2.1 5.0 16000 Manual 34903 Diesel 2.1 5.0 12500 Automatic 35345 Diesel 2.1 7.0 14495 Manual 23146 Petrol 2.0 6.0 24500 Automatic 16000 Diesel 2.1 4.0 23990 Automatic 44000 Petrol 6.2 8.0 17250 Automatic 30000 Diesel 2.1 6.0 14200 Automatic 22000 Petrol 1.6 6.0 22950 Automatic 19000 Petrol 2.0 4.0 10500 Automatic 90000 Diesel 2.1 8.0 15650 Automatic 32000 Diesel 2.1 7.0 14850 Automatic 90000 Diesel 2.1 6.0 22400 Automatic 37000 Diesel 2.1 4.0 19950 Automatic 23000 Diesel 2.1 4.0 9990 Automatic 78000 Diesel 2.1 7.0 11990 Automatic 47000 Diesel 2.1 6.0 20990 Automatic 18000 Diesel 2.1 4.0 14490 Automatic 64000 Diesel 2.1 6.0 10490 Automatic 50000 Diesel 2.1 8.0 15990 Automatic 65000 Diesel 2.1 5.0 17850 Automatic 38000 Diesel 2.1 5.0 16450 Automatic 70000 Diesel 2.1 5.0 11350 Automatic 69000 Diesel 2.1 8.0 12900 Automatic 41000 Petrol 1.6 7.0 26000 Automatic 16250 Petrol 1.5 3.0 22767 Automatic 30676 Diesel 2.1 4.0 12000 Automatic 40005 Diesel 2.1 6.0 18999 Automatic 38000 Hybrid 2.0 4.0 14000 Automatic 45000 Diesel 2.1 5.0 12000 Automatic 40005 Diesel 2.1 6.0 4343 Automatic 120000 Diesel 2.1 12.0 16500 Automatic 51936 Diesel 2.1 5.0 27750 Automatic 4000 Diesel 2.0 2.0 28950 Automatic 8500 Diesel 2.0 2.0 3495 Automatic 119168 Petrol 1.8 14.0 59740 Automatic 21800 Petrol 4.0 4.0 44990 Automatic 19999 Petrol 4.0 2.0 54990 Automatic 1600 Petrol 4.0 2.0 15995 Automatic 46221 Diesel 2.1 6.0 14000 Automatic 45000 Diesel 2.1 5.0 12000 Automatic 40005 Diesel 2.1 6.0 22767 Automatic 30676 Diesel 2.1 4.0 12998 Semi-Auto 52712 Diesel 2.1 7.0 17998 Semi-Auto 59858 Diesel 2.1 4.0 28998 Manual 22 Petrol 1.6 2.0 24998 Semi-Auto 28234 Petrol 2.0 4.0 19498 Semi-Auto 22803 Diesel 2.1 5.0 23498 Semi-Auto 17093 Diesel 2.1 3.0 25498 Semi-Auto 8150 Diesel 1.6 2.0 24998 Semi-Auto 9344 Diesel 2.0 2.0 15498 Semi-Auto 36243 Diesel 2.1 7.0 20995 Semi-Auto 33000 Diesel 2.1 4.0 31995 Semi-Auto 11300 Petrol 3.0 3.0 23995 Semi-Auto 32800 Diesel 2.1 4.0 16995 Automatic 37600 Diesel 2.1 7.0 24995 Automatic 14000 Petrol 2.0 4.0 26998 Semi-Auto 24175 Petrol 6.2 7.0 20498 Semi-Auto 27968 Diesel 2.1 5.0 27498 Semi-Auto 9255 Diesel 2.0 3.0 16998 Semi-Auto 26973 Diesel 1.6 5.0 27498 Semi-Auto 12491 Diesel 2.1 3.0 27698 Semi-Auto 14999 Diesel 2.0 2.0 27498 Semi-Auto 37941 Petrol 3.0 5.0 32498 Semi-Auto 11301 Diesel 2.0 2.0 24498 Semi-Auto 12629 Diesel 2.1 2.0 21498 Semi-Auto 20981 Diesel 2.1 4.0 18998 Semi-Auto 23760 Diesel 2.1 4.0 

## 9. Summary

Test & Hypothesis

In order to understand what affects change in price of a used car, the relation between features available in the data sat will be examined by using inferential statistic methods. The primary assumption based on figures and tables is price must be affected by odometer and condition. There must be other features that affects price significantly. It will be investigated in the later phase of the study.

The first hypothesis:

Null Hypothesis: There is no significant relation between price and odometer of a car

Alt-Hypothesis: There is a significant relation between price and odometer.

The second hypothesis:

Null Hypothesis: There is no significant relation between price and condition of a car

Alt-Hypothesis: There is a significant relation between price and condition.

Odometer vs Price

Firstly, examine first hypothesis. Independent t-test is an appropriate method while examining the relation between two numerical variables. On the other hand, this test has some presumptions. Homogeneity of variances is one of the assumptions. In order to check whether homogeneity of variance assumption is violated or not, Levene test is applied.

## Check next notebook for ML Model: 

This section used applied machine learning models as a framework for the data analysis. The data set is a supervised data which refers to fitting a model of dependent variables to the independent variables, with the goal of accurately predicting the dependent variable for future observations or understanding the relationship between the variables.

- Pre-processing the data

- Label Encoding. In the dataset, there are 6 predictors. 3 of them are numerical variables while rest of them are categorical. In order to apply machine learning models, we need numeric representation of the features. Therefore, all non-numeric features were transformed into numerical form.

- Train the data. In this process, 20% of the data was split for the test data and 80% of the data was taken as train data.

- Scaling the Data. While exploring the data in the previous sections, it was seen that the data is not normally distributed. Without scaling, the machine learning models will try to disregard coefficients of features that has low values because their impact will be so small compared to the big value features.

**[Daimler]-DataScience_aptituteTest_MarcoRusso_v1.1.0_[part II]_python-scala**

## Version Control

- 1.2.0 - MR - 2021.07.20
```
AddDelta Table
Save Parquet Table
Remove check parquet table
```

- 1.1.0 
Original notebook